demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,483 @@
+                                                     topicFilterLength,
+                                                     topicNameLength,
+                                                     nameIndex,
+                                                     filterIndex );
+        }
+        else
+        {
+            /* Check for matching wildcards. */
+            shouldStopMatching = matchWildcards( pTopicFilter,
+                                                 pTopicName,
+                                                 topicNameLength,
+                                                 filterIndex,
+                                                 &nameIndex,
+                                                 &matchFound );
+        }
+
+        if( ( matchFound == true ) || ( shouldStopMatching == true ) )
+        {
+            break;
+        }
+
+        /* Increment indexes. */
+        nameIndex++;
+        filterIndex++;
+    }
+
+    if( matchFound == false )
+    {
+        /* If the end of both strings has been reached, they match. This represents the
+         * case when the topic filter contains the '+' wildcard at a non-starting position.
+         * For example, when matching either of "sport/+/player" OR "sport/hockey/+" topic
+         * filters with "sport/hockey/player" topic name. */
+        matchFound = ( ( nameIndex == topicNameLength ) &&
+                       ( filterIndex == topicFilterLength ) ) ? true : false;
+    }
+
+    return matchFound;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool matchTopic( const char * pTopicName,
+                        const uint16_t topicNameLength,
+                        const char * pTopicFilter,
+                        const uint16_t topicFilterLength )
+{
+    assert( pTopicName != NULL );
+    assert( topicNameLength != 0 );
+    assert( pTopicFilter != NULL );
+    assert( topicFilterLength != 0 );
+
+    bool status = false;
+    bool topicFilterStartsWithWildcard = false;
+
+    /* Check for an exact match if the incoming topic name and the registered
+     * topic filter length match. */
+    if( topicNameLength == topicFilterLength )
+    {
+        status = ( strncmp( pTopicName, pTopicFilter, topicNameLength ) == 0 ) ? true : false;
+    }
+
+    /* Determine if topic filters starts with a wildcard. */
+    topicFilterStartsWithWildcard = ( ( pTopicFilter[ 0 ] == '+' ) ||
+                                      ( pTopicFilter[ 0 ] == '#' ) ) ? true : false;
+
+    /* If an exact match was not found, match against wildcard characters in topic filter.
+     * Note: According to the MQTT 3.1.1 specification, incoming PUBLISH topic names starting
+     * the "$" character cannot be matched against topic filter starting with a wildcard.
+     * i.e. for example, "$SYS/sport" cannot be matched with "#" or "+/sport" topic filters. */
+    if( ( status == false ) && !( ( ( pTopicName[ 0 ] == '$' ) ) &&
```

Comment:
Too many parentheses?
```suggestion
    if( ( status == false ) && !( ( pTopicName[ 0 ] == '$' ) &&
```